### PR TITLE
Fix Authentication Request Retries

### PR DESCRIPTION
**Pull request by PJSimon**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @PJSimon has signed the CLA. Thanks for the pull request!

File: Box.V2/JWTAuth/BoxJWTAuth.cs [hunk 1]
```diff
@@ -145,27 +149,61 @@ public string UserToken(string userId)
...
+                        ||
+                        (ex.StatusCode == HttpStatusCode.Accepted && retryAfterHeader != null))
+                        && retryCounter++ < HttpRequestHandler.RetryLimit)
+                    {
+                        // Before we retry the JWT Authentication request, we must regenerate the JTI claim with an updated datetime.
+                        var serverDate = ex.ResponseHeaders != null ? ex.ResponseHeaders.Date : null;
+                        if (serverDate.HasValue)
+                        {
+                            var date = serverDate.Value;
+                            assertion = ConstructJWTAssertion(subId, subType, date.LocalDateTime);
```
Comment (sujaygarlanka) [inline Box.V2/JWTAuth/BoxJWTAuth.cs]: Do we want to add the delay to the expiration time?

File: Box.V2/JWTAuth/BoxJWTAuth.cs [hunk 2]
```diff
@@ -145,27 +149,61 @@ public string UserToken(string userId)
...
+                        }
+                        else
+                        {
+                            assertion = ConstructJWTAssertion(subId, subType, DateTime.Now);
+                        }
+
+                        TimeSpan delay = expBackoff.GetRetryTimeout(retryCounter);
```
Comment (sujaygarlanka) [inline Box.V2/JWTAuth/BoxJWTAuth.cs]: I think the retry-after header time should be used here for the delay if available.

File: Box.V2/JWTAuth/BoxJWTAuth.cs [hunk 3]
```diff
@@ -145,27 +149,61 @@ public string UserToken(string userId)
...
+                    if ((ex.StatusCode == HttpRequestHandler.TooManyRequests
+                        ||
+                        ex.StatusCode == HttpStatusCode.InternalServerError
+                        ||
+                        ex.StatusCode == HttpStatusCode.BadGateway
+                        ||
+                        ex.StatusCode == HttpStatusCode.ServiceUnavailable
+                        ||
+                        ex.StatusCode == HttpStatusCode.GatewayTimeout
+                        ||
```
Comment (sujaygarlanka) [inline Box.V2/JWTAuth/BoxJWTAuth.cs]: May want to retry on the 400 error where the Unix time on the user's machine does not match the server.

File: Box.V2/JWTAuth/BoxJWTAuth.cs (hunk 1 above)
Comment (PJSimon) [reply Box.V2/JWTAuth/BoxJWTAuth.cs]: Done!

File: Box.V2/JWTAuth/BoxJWTAuth.cs (hunk 2 above)
Comment (PJSimon) [reply Box.V2/JWTAuth/BoxJWTAuth.cs]: Done!

File: Box.V2/JWTAuth/BoxJWTAuth.cs (hunk 3 above)
Comment (PJSimon) [reply Box.V2/JWTAuth/BoxJWTAuth.cs]: Done!

**Pull request merged**